lib/compat.php

```diff
@@ -77,6 +77,69 @@ function gutenberg_provide_render_callback_with_block_object( $pre_render, $bloc
 }
 add_filter( 'pre_render_block', 'gutenberg_provide_render_callback_with_block_object', 10, 2 );
 
+/**
+ * Registers the user meta property responsible for storing preferences for the
+ * data script persistence middleware. The meta is explicitly registered so that
+ * it is made accessible for read and write via the REST API.
+ */
+function gutenberg_register_data_persistence_user_meta() {
+	register_meta( 'user', 'wp_data_persistence', array(
+		'type'         => 'string',
+		'single'       => true,
+		'show_in_rest' => true,
+	) );
+}
+add_action( 'init', 'gutenberg_register_data_persistence_user_meta' );
+
+/**
+ * Enqueues inline scripts used to provide a custom storage interface for the
+ * data script persistence middleware. The custom storage implementation stores
+ * persisted values using user meta, bootstrapped for initial read at load-time,
+ * and updated via the REST API on change.
+ */
+function gutenberg_user_settings_data_persistence_inline_script() {
+	global $wp_scripts;
+
+	$user_id         = get_current_user_id();
+	$persisted_value = get_user_meta( $user_id, 'wp_data_persistence', true );
+	if ( empty( $persisted_value ) ) {
+		// If there's no explicit metadata assigned, fall back to a value which
+		// was persisted using browser storage, prior to user meta persistence.
+		$persisted_value = sprintf( 'localStorage.getItem( "WP_DATA_USER_%s" );', $user_id );
+	} else {
+		// Otherwise, encode the string value for interpolation in the storage
+		// implementation script.
+		$persisted_value = json_encode( $persisted_value );
+	}
+
+	$persistence_script = <<<JS
+( function() {
+	wp.data.use( wp.data.plugins.persistence, {
+		storage: {
+			getItem: function() {
+				return {$persisted_value};
+			},
+			setItem: function( key, value ) {
+				wp.apiFetch( {
+					path: '/wp/v2/users/me',
+					method: 'POST',
+					data: {
+						meta: {
+							wp_data_persistence: value,
+						}
+					}
+				} );
+			}
+		}
+	} );
+} )();
+JS;
+
+	$wp_scripts->registered['wp-data']->extra['after'] = array();
+	wp_add_inline_script( 'wp-data', $persistence_script );
+}
+add_action( 'enqueue_block_editor_assets', 'gutenberg_user_settings_data_persistence_inline_script', 20 );
```

Comment:
Note: The priority here shouldn't be necessary, but the behavior of the function is such that it replaces any inline scripts for the `wp-data` handle. This conflicts with Gutenberg's own [`gutenberg_enqueue_block_editor_assets` function](https://github.com/WordPress/gutenberg/blob/ab8af23410f9f89f5a57064b43758b157f392d55/lib/client-assets.php#L423), which does the same and was intended to serve as a temporary implementation of the preferences migration. Since this was migrated to core as part of [Trac#46429](https://core.trac.wordpress.org/ticket/46429), it should be removed from Gutenberg, at which point the priority here would no longer be necessary.

Reply:
>Since this was migrated to core as part of [Trac#46429](https://core.trac.wordpress.org/ticket/46429), it should be removed from Gutenberg, at which point the priority here would no longer be necessary.

See #19178 . If the other is merged first, this can be rebased to remove the priority parameter.